### DISCUSSION TITLE
A trap of `switch` pattern

### DISCUSSION THREAD

**Discussion by LeaFrock** [dotnet/csharplang#8124] [Show and tell]
Without debugging, tell me the result of `x`.

```c#
int c = 3;

int x = c % 3 switch
{
    0 => c / 3,
    1 => (c - 4) / 3 + 2,
    _ => (c - 2) / 3 + 1,
};

```

**Comment by KennethHoff** [comment]
I'd imagine it'd be `1`, but I know that the `%` operator is kind of weird.

**Comment by HaloFour** [comment]
The result is `0`.

Seems to be an issue of precedence, `switch` is evaluated before `%`.

**Reply by HaloFour** [reply]
Here's the documentation on operator precedence:

https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/#operator-precedence

**Comment by kyoyama-kazusa** [comment]
You should treat `switch` as a unary expression so this expression will be treated as follows:

```csharp
int c = 3;
int x = c % (3 switch
{
    0 => c / 3,
    1 => (c - 4) / 3 + 2,
    _ => (c - 2) / 3 + 1,
});
```

`switch` expression is an "entire" expression, meaning you cannot split a `switch` expression into two individual parts (`3` and `{ ... }`).

Therefore the value is 0 (`3 switch { _ => (c - 2) / 3 + 1 }` -> `1 / 3 + 1` -> 1, `3 % 1` -> 0).

---

There's a potential rule for operators: The precedence of binary expressions is lower than unary expressions, which means you should firstly calculate for the result of an unary, and then binary if they are mixed without braces.

From this rule, the expression won't be considered as `(c % 3) switch`, because `c % 3` is a binary expression, which is lower than `3 switch {...}`.

> Here is just one exception: range operator `x..y` is before unary expressions `switch` and `with`. This is because the type of two operands must be an `int` value or an expression that can be implicitly converted into an `int` value. This rule limits the usage on operands.
> In addition, operands in range operators can be omitted. This is the main reason why it is before all binary operators and we should add an extra brace if we use `+`, `-` and other binary operators.

**Reply by LeaFrock** [reply]
A great 'Show and tell' 👍

**Comment by RasheedHendawi** [comment]
### Switch will work first *
because the precedence of unary expressions is higher than binary.
